File: JavaNgArch/src/javangarch/mvc/gn/UserController.java
package javangarch.mvc.gn;
import javax.annotation.Resource;
import javax.validation.Valid;

import org.apache.commons.lang3.StringUtils;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import projectbase.domain.DORef;
import projectbase.domain.ICommonBD;
import projectbase.domain.customcollection.DomainList;

import projectbase.mvc.Auth;
import projectbase.mvc.AuthFailureException;
import projectbase.mvc.Transaction;
import projectbase.mvc.result.ActionResult;
import projectbase.utils.HqlCriterion;
import projectbase.utils.Util;
import javangarch.domain.domainmodel.gn.*;
import javangarch.mvc.shared.AppBaseController;


    @Auth
    public class UserController extends AppBaseController
    {
        @Resource public ICommonBD<User> UserBD;
        @Resource public ICommonBD<Dept> DeptBD;
        @Resource public ICommonBD<Func> FuncBD;

        public ActionResult Search()
        {
        	UserSearchVM m=new UserSearchVM();
        	HqlCriterion filter = new HqlCriterion().And("this.corp.id = ?",GetLoginCorpId());
        	m.setDepts(DeptBD.GetRefList(filter, "Code"));
            return ForView(m);
        }
        
        public ActionResult List(@Valid UserSearchVM searchvm,UserListVM listvm)
        {
        	HqlCriterion filter = new HqlCriterion().And("this.corp.id = ?",GetLoginCorpId());
            if (Util.IsNotNull(searchvm.getInput().getDept()))
            {
                listvm.setDeptId(searchvm.getInput().getDept().getId());
                filter = new HqlCriterion().And("this.dept.id = ?",listvm.getDeptId());
            }
            listvm.setResultList(UserBD.GetDtoList(listvm.getInput().getPager(), 
            		UserListVM_ListRow.class,filter,listvm.getInput().getOrderExpression()));
            return ForView("List",listvm);
        }

        @RequestMapping(method=RequestMethod.POST)
        @Transaction
        public ActionResult Delete(UserSearchVM search,UserListVM list)
        {
            for (Integer id : list.getInput().getSelectedValues())
                UserBD.Delete(id);
            return List(search,list);
        }
        public ActionResult Add(Integer deptId)
        {
        	UserEditVM m = new UserEditVM();
        	m.getInput().setIsActive(true);
            if (deptId != null) m.getInput().setDept(new DORef<Dept>(deptId.intValue()));
            m.setCanChangeUserFunc(CanAccess("_ChangeUserFunc"));
            return ForView("Edit",m);
        }

        public ActionResult Edit(int id)
        {
            User user = UserBD.Get(id);
            UserEditVM m = new UserEditVM();
            Map(user,m.getInput());

            for (Func func : user.getDept().getCorp().getFuncs())
            {
                m.getInput().setCorpFuncIds(m.getInput().getCorpFuncIds() + "," + func.getId());
            }
            if (!m.getInput().getCorpFuncIds().isEmpty()) m.getInput().setCorpFuncIds(m.getInput().getCorpFuncIds().substring(1));
            for (Func func : user.getDept().getFuncs())
            {
                m.getInput().setDeptFuncIds(m.getInput().getDeptFuncIds() + "," + func.getId());
            }
            if (!m.getInput().getDeptFuncIds().isEmpty()) m.getInput().setDeptFuncIds(m.getInput().getDeptFuncIds().substring(1));
            for (Func func : user.getFuncs())
            {
                m.getInput().setUserFuncIds(m.getInput().getUserFuncIds() + "," + func.getId());
            }
            if (!m.getInput().getUserFuncIds().isEmpty()) m.getInput().setUserFuncIds(m.getInput().getUserFuncIds().substring(1));

            m.setCanChangeUserFunc (!user.equals(GetLoginUser()) && CanAccess("_ChangeUserFunc"));
            return ForView(m);
        }

        @RequestMapping(method=RequestMethod.POST)
        @Transaction
        public ActionResult Save(@Valid UserEditVM editvm) 
        {
            Save(editvm.getInput());
            if (editvm.getInput().getId() == 0) return RcJson();
            return ClientShowMessage(Message_SaveSuccessfully);
        }
        @RequestMapping(method=RequestMethod.POST)
        @Transaction
        public ActionResult SavePassword(String OldPassword, String NewPassword)
        {
                User user = GetLoginUser();
                if (!user.getPassword().equals(OldPassword))
                {
                    SetViewMessage("OldPassword_Wrong");
                }
                else
                {
                    //user.setPassword ( NewPassword);
                    SetViewMessage(Message_SaveSuccessfully);
                }

            return ClientShowMessage();
        }
        @RequestMapping(method=RequestMethod.POST)
        @Transaction
        public ActionResult ResetPassword(int userId)
        {
            if (!IsAdmin()) return AuthFailure();
            User user = UserBD.Get(userId);
            user.setPassword( DefaultPassword);

            return ClientShowMessage("ResetPassword_Succeed");
        }
  
        @Override
        protected  void OnViewExecuting(Object viewModel)
        {
        	if(viewModel instanceof UserEditVM)
        		((UserEditVM)viewModel).setDepts(DeptBD.GetRefList(new HqlCriterion().And("this.corp.id =?", GetLoginCorpId())));
        }
        
         private User Save(UserEditVM.EditInput input) 
        {
            User user;
            if (input.getId() == 0)
            {
                user = new User();
                user.setCorp(GetLoginCorp());
                user.setPassword(DefaultPassword);
                user.setFuncs( new DomainList<Func>());
            }
            else
            {
                user = UserBD.Get(input.getId());
            }
            if (IsAdmin(user)) throw new AuthFailureException();
			user.setDept ( input.getDept().ToReferencedDO(DeptBD));
            if (!AdminCode.equalsIgnoreCase(user.getCode()))
            {
                user.setCode(input.getCode());
            }
            user.setName(input.getName());
			user.setCellPhone(input.getCellPhone());
            user.setIsActive(input.getIsActive());
            if (!user.equals(GetLoginUser()))
            {
                if (CanAccess("_ChangeUserFunc"))
                {
                    user.getFuncs().clear();
                    if (!StringUtils.isEmpty(input.getUserFuncIds()))
                    {
                    	for(String id:input.getUserFuncIds().split(","))
                        	user.getFuncs().add(FuncBD.Get(Integer.valueOf(id)));
                    }
                }
            }
            UserBD.Save(user);
            return user;
        }

    }

File: NetNgArch/src/NetNgArch.Web.Mvc/GN/UserController.cs
using System;
using System.Collections.Generic;
using System.Web.Mvc;
using NetNgArch.Domain.DomainModel.GN;
using NetNgArch.Web.Mvc.Shared;
using ProjectBase.Domain;
using ProjectBase.Utils;
using ProjectBase.Web.Mvc;
using System.Linq.Expressions;


namespace NetNgArch.Web.Mvc.GN
{
    [Auth]
    public class UserController : AppBaseController
    {
        public ICommonBD<User> UserBD { get; set; }
        public ICommonBD<Dept> DeptBD { get; set; }
        public ICommonBD<Func> FuncBD { get; set; }

        public ActionResult Search()
        {
            return ForView(new UserSearchVM()
                               {
                                   Depts = DeptBD.GetRefList(o => o.Corp.Id == GetLoginCorpId())
                               });
        }

        public ActionResult List(UserSearchVM searchvm,UserListVM listvm)
        {
            Expression<Func<User,bool>> filter = o => o.Corp.Id == GetLoginCorpId();
            if (Util.IsNotNull(searchvm.Input.Dept))
            {
                listvm.DeptId = searchvm.Input.Dept.Id;
                filter = o => o.Dept.Id == searchvm.Input.Dept.Id;
            }
            listvm.ResultList = UserBD.GetDtoList<UserListVM.ListRow>(listvm.Input.Pager, filter, listvm.Input.OrderExpression);
            return ForView("List",listvm);
        }

        [HttpPost]
        [Transaction]
        public ActionResult Delete(UserSearchVM searchvm, UserListVM listvm)
        {
            foreach (var id in listvm.Input.SelectedValues)
                UserBD.Delete(id);
            return List(searchvm, listvm);
        }
        public ActionResult Add(int? deptId)
        {
            var m = new UserEditVM {Input = new UserEditVM.EditInput {IsActive = true}};
            if (deptId != null) m.Input.Dept = new DORef<Dept>(deptId.Value);
            m.CanChangeUserFunc=CanAccess("_ChangeUserFunc");
            return ForView("Edit",m);
        }

        public ActionResult Edit(int id)
        {
            var user = UserBD.Get(id);
            var m = new UserEditVM
            {
                Input = Map<User, UserEditVM.EditInput>(user)
            };
            foreach (var func in user.Dept.Corp.Funcs)
            {
                m.Input.CorpFuncIds = m.Input.CorpFuncIds + "," + func.Id;
            }
            if (m.Input.CorpFuncIds != "") m.Input.CorpFuncIds = m.Input.CorpFuncIds.Substring(1);
            foreach (var func in user.Dept.Funcs)
            {
                m.Input.DeptFuncIds = m.Input.DeptFuncIds + "," + func.Id;
            }
            if (m.Input.DeptFuncIds != "") m.Input.DeptFuncIds = m.Input.DeptFuncIds.Substring(1);
            foreach (var func in user.Funcs)
            {
                m.Input.UserFuncIds = m.Input.UserFuncIds + "," + func.Id;
            }
            if (m.Input.UserFuncIds != "") m.Input.UserFuncIds = m.Input.UserFuncIds.Substring(1);
            m.CanChangeUserFunc=(!user.Equals(GetLoginUser()) && CanAccess("_ChangeUserFunc"));
            return ForView(m);
        }

        [HttpPost]
        [Transaction]
        public ActionResult Save(UserEditVM editvm)
        {
            if (ModelState.IsValid)
            {
                Save(editvm.Input);
                SetViewMessage(Message_SaveSuccessfully);
                if (editvm.Input.Id == 0) return RcJson();
            }
            return ClientShowMessage();
        }
        [HttpPost]
        [Transaction]
        public ActionResult SavePassword(string oldpassword, string newpassword)
        {
            if (!ModelState.IsValid)
                SetViewMessage(Message_UserInputError);
            else
            {
                var user = GetLoginUser();
                if (user.Password != oldpassword)
                {
                    SetViewMessage(Res.M("OldPassword_Wrong"));
                }
                else
                {
                    user.Password = newpassword;
                    SetViewMessage(Message_SaveSuccessfully);
                }
            }
            return ClientShowMessage();
        }
        [HttpPost]
        [Transaction]
        public ActionResult ResetPassword(int userId)
        {
            if (!IsAdmin()) return AuthFailure();
            var user = UserBD.Get(userId);
            user.Password = DefaultPassword;

            return ClientShowMessage(Res.M("ResetPassword_Succeed"));
        }
      
        protected override void OnViewExecuting(object viewModel)
        {
            var l = DeptBD.GetRefList(o => o.Corp.Id == GetLoginCorpId());
            SetViewModel<UserEditVM>(m => m.Depts =l );
            SetViewModel<UserSearchVM>(m => m.Depts =l );
        }
        private void Save(UserEditVM.EditInput input)
        {
            User user;
            if (input.Id == 0)
            {
                user = new User();
                user.Corp = GetLoginCorp();
                user.Password = DefaultPassword;
                user.Funcs = new DomainList<Func>();
            }
            else
            {
                user = UserBD.Get(input.Id);
            }
            if (IsAdmin(user)) throw new AuthFailureException();
			user.Dept = input.Dept.ToReferencedDO(DeptBD);
            if (!AdminCode.Equals(user.Code, StringComparison.OrdinalIgnoreCase))
            {
                user.Code = input.Code;
            }
            user.Name = input.Name;
            user.IsActive = input.IsActive;
            if (user!=GetLoginUser())
            {
                if (CanAccess("_ChangeUserFunc"))
                {
                    user.Funcs.Clear();
                    if (!string.IsNullOrEmpty(input.UserFuncIds))
                    {
                        Array.ForEach(input.UserFuncIds.Split(','), o => user.Funcs.Add(FuncBD.Get(int.Parse(o))));
                    }
                }
            }
            UserBD.Save(user);
        }

    }



}
